### PR TITLE
Initial persist-last; clean up channel namespace docs

### PR DESCRIPTION
**Pull request by SimonWoolf**
[DOC-13]

channel namespaces docs: https://github.com/ably/docs/issues/887

persist-last: https://ably.atlassian.net/browse/DOC-13

[DOC-13]: https://ably.atlassian.net/browse/DOC-13

### CODE REVIEW

**Comment by MarkWoulfeAbly** [comment]
Thanks @SimonWoolf. Just a few questions and I'll also loop in @JackUkleja from product and @tomczoink from documentation standpoints if their input is needed.

1. The channel rules and namespaces page that you updated with the partial is currently listed as a 'REST API beta feature' and not listed in the right-hand navigation/TOC - https://docs.ably.io/general/channel-rules-namespaces/. Given the partial is appearing on other 'live' pages I'm guessing it does not make sense to be a hidden page anymore? Also is it REST only as the beta title would suggest?

1. Is the channel-param entry also relevant for v1.2 and onwards? If so I can add it to the page the new page we made for that in v1.2. Which I can now see may be confusing and might need some redirects in a separate PR.

**Comment by SimonWoolf** [comment]
>  is currently listed as a 'REST API beta feature' 

I'm not clear on what the mechanism is for it be 'listed as a rest api beta feature' at the moment. I can't see anything in the textile file marking it as such. But in any case, no, rules/namespaces is not and has never been a beta feature.

> I'm guessing it does not make sense to be a hidden page anymore? 

See the discussion on the other issue, which I linked. Ivan's proposal on https://github.com/ably/docs/issues/887#issuecomment-646195655 , which I agree with, is to deprecate the [solutions article](https://support.ably.io/support/solutions/articles/3000030058-what-is-a-channel-namespace-and-how-can-i-use-them-) (replace with a link to the main docs page), and make the docs page more prominent.

> Is the channel-param entry also relevant for v1.2 and onwards

Yes -- sorry, my mistake to add it only to the v1.1 page.

**Comment by MarkWoulfeAbly** [comment]
Thanks for clarifying @SimonWoolf. @tomczoink are you aware why/how this page has become marked as a REST beta feature?

**Comment by tomczoink** [comment]
I've pushed a fix to the master branch to fix that, wasn't related to the changes in this PR.

**Comment by MarkWoulfeAbly** [comment]
I've cherry-picked @tomczoink's changes, tweaked the copy slightly and added it to the v1.2 docs.

**Comment by JackUkleja** [comment]
I'm still fairly disturbed by our inconsistent/incoherent terminology around channel names, channels IDs, and namespaces, although I'm not sure it's appropriate to tackle as part of this PR.

The namespace one in particular is relevant here. Channel names are defined [here](https://support.ably.com/support/solutions/articles/3000046824-what-restrictions-are-there-on-channel-names-). Channels names optionally include a namespace. But for Channel Rules we consider any <text you put in the box> without a namespace prefix to be a namespace (not a non-namespaced channel name). The <text you put in the box> is also labelled "Namespace or Channel ID"

**Comment by JackUkleja** [comment]
Proposal to change terminology as follows:
* We change _Channel rules_ to _Channel settings_. 
  * When we talk about rules in the context of _Channel settings_ we should just use the word _rules_ (this matches the dashboard UI which says "Add new rule")

This PR should reflect that. Another PR will be created to make relevant changes to dashboard. @MarkWoulfeAbly to coordinate with CS about any changes needed in KB articles

**Comment by SimonWoolf** [comment]
> We change Channel rules to Channel settings. 

I endorse this 👍. 'Rules' as a term is too overloaded with reactor rules.

**Comment by JackUkleja** [comment]
@SimonWoolf can you help with an answer to https://github.com/ably/docs/pull/944#issuecomment-685872040 
Alternatively slack or call me/Mark if it's quicker

**Comment by SimonWoolf** [comment]
> Is my understanding below correct?
> * `public` (namespace = `public`, channelName= `null`, validChannelID = `false`)
> * `public:events` (namespace = `public`, channelName= `events`, validChannelID = `true`)
> * `public:news:americas` (namespace = `public`, channelName= `news:americas`, validChannelID = `true`)
> Am I right in thinking then you can't have a channel name without a namespace?

No. The channel name doesn't exclude the namespace. The namespace is just the part of the channel name before the first colon. Correct version would be:
* `public` (namespace = `public`, channelName= `public`)
* `public:events` (namespace = `public`, channelName= `public:events`)
* `public:news:americas` (namespace = `public`, channelName= `public:news:americas`)

I don't know what you mean by "validChannelID".

> Is channel ID even a discrete concept?

At least internally in realtime, the distinction between channelName and channelId is that the id includes the scope and qualifier, if either is present, the name excludes them e.g. a channel with id `[product:foo/bar]baz` has channel name `baz`; a channel with channel id `[meta]log` has channel name `log`. For channels with neither scope nor qualifier the two are the same. (And client libraries do not make the distinction because the channel is opaque to client libs)

Of course we don't have to follow that naming convention in public docs if we don't think it's a useful distinction for a customer.

**Comment by JackUkleja** [comment]
> > Is channel ID even a discrete concept?
> 
> At least internally in realtime, the distinction between channelName and channelId is that the id includes the scope and qualifier, if either is present, the name excludes them e.g. a channel with id `[product:foo/bar]baz` has channel name `baz`; a channel with channel id `[meta]log` has channel name `log`. For channels with neither scope nor qualifier the two are the same. (And client libraries do not make the distinction because the channel is opaque to client libs)

So when the "Channel Rule" UI talks about "Namespace or Channel ID" is it saying that you can provide:

* `public` (a namespace AKA channel name)
* `[scope:qualfier]public` (a channel ID)

I find it odd we refer to the latter as a "Channel ID" not a namespace

**Comment by SimonWoolf** [comment]
> So when the "Channel Rule" UI talks about "Namespace or Channel ID" is it saying that you can provide:
> - public (a namespace AKA channel name)
> - [scope:qualfier]public (a channel ID)

I don't think so -- I think the field being called "namespace or channel id" was just an attempt to gesture at the fact that if the channel name has no colon, the namespace is just the whole channel id. It's quite a confusing guesture, and the field should probably be renamed (eg to just "namespace").

It was named that since years before we had metachannels or api-streamer or anything else, back then the square-brackets scope syntax just wasn't exposed to customers anywhere, and 'channel id' and 'channel name' were I think used basically interchangeably in the documentation. That distinction I mentioned is certainly not consistently adhered to in public docs in the same way it is in the internal codebase.

To be clear: if you want a channel rule to apply to, say, a metachannel, you do need to include the qualifier. A rule for `log` will not apply to, say, `[meta]log:push`, you need `[meta]log`. (It would be quite confusing if someone made a rule for their own `log:foo` channels and suddenly found it applied to `[meta]log` as well, which after all lives in a whole different... uh, namespace. .Not that metachannels obey user-configured persistence settings at the moment)

**Comment by MarkWoulfeAbly** [comment]
@SimonWoolf in the case where there is no colon, is it both a channel name and a namespace you're providing? Are they considered the same thing at that point?

I'm trying to think is there a use case where you want to specify a channel name but not a namespace or vice versa.

**Comment by JackUkleja** [comment]
> @SimonWoolf in the case where there is no colon, is it both a channel name and a namespace you're providing?
 Are they considered the same thing at that point?

Yes, see https://github.com/ably/docs/pull/944#issuecomment-686425312

> I'm trying to think is there a use case where you want to specify a channel name but not a namespace or vice versa.

A channel name always includes a namespace, so there is no way to specify it without. 

Providing just a namespace will match all channel names that are only namespace, plus any channel name _within_ the namespace.

Yes, this is a little confusing

File: content/partials/shared/_channel_namespaces.textile [hunk 1]
```diff
@@ -1,11 +1,17 @@
...
-* **Require identification** - if enabled, clients will not be permitted to subscribe to matching channels unless they are both authenticated and identified (they have an assigned client ID). Anonymous clients are not permitted to join these channels. Find out more about "authenticated and identified clients":https://support.ably.com/solution/articles/3000038608-authenticated-clients
-* **Require TLS** - if enabled, only clients who have connected to Ably over TLS will be allowed to join the channel
+The namespace attributes that can be configured are:
 
-Key or token capabilities can also specify access rights based on channel namespace, find out more about "authentication":/core-features/authentication
+* **Persisted** - if enabled, all messages within this namespace will be stored according to the storage rules for your account (24 hours for free accounts). You can access stored messages via the "history API":/realtime/history . Please note that for each message stored, an additional message is deducted from your monthly allocation.
```
[the remark this answers is not in the document]
Comment (kavalerov) [reply content/partials/shared/_channel_namespaces.textile:11]: +1

File: content/partials/shared/_channel_namespaces.textile [hunk 2]
```diff
@@ -1,11 +1,17 @@
...
+* **Persist last** - if enabled, the very last message published on a channel will be stored for an entire year, retrievable using the "channel rewind mechanism":https://www.ably.io/blog/channel-rewind/ by attaching to the channel with "rewind=1.":/realtime/channel-params#rewind Only messages are stored, not presence messages. This last message storage is not accessible using the normal history API, only through rewind.
```
[the remark this answers is not in the document]
Comment (kavalerov) [reply content/partials/shared/_channel_namespaces.textile:12]: +1

File: content/partials/shared/_channel_namespaces.textile [hunk 3]
```diff
@@ -1,11 +1,17 @@
...
+* **Identified** - if enabled, clients will not be permitted to use (including to attach, publish, or subscribe) matching channels unless they are "identified":/core-features/authentication#identified-clients (they have an assigned client ID). Anonymous clients are not permitted to join these channels. Find out more about "authenticated and identified clients":https://support.ably.com/solution/articles/3000038608
+* **TLS Only** - if enabled, only clients who have connected to Ably over TLS will be allowed to use matching channels. "Find out more":https://support.ably.com/solution/articles/3000045208
```
[the remark this answers is not in the document]
Comment (kavalerov) [reply content/partials/shared/_channel_namespaces.textile:14]: +1

File: content/realtime/channels/channel-parameters/rewind.textile [hunk 4]
```diff
@@ -132,6 +132,10 @@ The channel position expressed by a @rewind@ parameter only has an effect on an
 
 Any @rewind@ parameter value that cannot be parsed either as a number or a time specifier will cause the attachment request to fail and return an error.
 
+h3(#persist-last-message). Persist last message
+
+If you have enabled the "**Persist Last** channel namespace rule,":/general/channel-rules-namespaces/ on the channel in question, you can attach with @rewind=1@ to retrieve the last message. Note that only the last message can be stored long-term (up to a year), and persisting the last message does not work for presence messages.
```
[the remark this answers is not in the document]
Comment (kavalerov) [reply content/realtime/channels/channel-parameters/rewind.textile:137]: +1

File: content/realtime/versions/v1.1/channel-params.textile [hunk 5]
```diff
@@ -89,6 +89,8 @@ At most 100 messages will be sent in a rewind request. If the number of messages
 
 By default, a maximum of two minutes of channel history is available when attaching. This means that a rewind time specifier of greater than two minutes will only be able to rewind by two minutes. If a channel has persistence enabled, then it is possible to rewind back in time by up to the persistence TTL on the channel.
 
+If you have enabled the "**Persist Last** channel namespace rule,":/general/channel-rules-namespaces/ on the channel in question, you can attach with @rewind=1@ to retrieve the last message. Note that only the last message can be stored long-term (up to a year), and persisting the last message does not work for presence messages.
```
[the remark this answers is not in the document]
Comment (kavalerov) [reply content/realtime/versions/v1.1/channel-params.textile:92]: +1

**Review by kavalerov — CHANGES_REQUESTED** [review]
Jacks comments are on point, so require fixing

File: content/partials/shared/_channel_namespaces.textile (hunk 1 above)
[the remark this answers is not in the document]
Comment (MarkWoulfeAbly) [reply content/partials/shared/_channel_namespaces.textile:11]:
```suggestion
* **Persist all messages** - if enabled, all messages within this namespace will be stored according to the storage rules for your account (24 hours for free accounts). You can access stored messages via the "history API":/realtime/history . Please note that for each message stored, an additional message is deducted from your monthly allocation.
```

File: content/partials/shared/_channel_namespaces.textile (hunk 2 above)
[the remark this answers is not in the document]
Comment (MarkWoulfeAbly) [reply content/partials/shared/_channel_namespaces.textile:12]:
```suggestion
* **Persist last message** - if enabled, the very last message published on a channel will be stored for an entire year, retrievable using the "channel rewind mechanism":https://www.ably.io/blog/channel-rewind/ by attaching to the channel with "rewind=1.":/realtime/channel-params#rewind Only messages are stored, not presence messages. This last message storage is not accessible using the normal history API, only through rewind.
```

File: content/partials/shared/_channel_namespaces.textile (hunk 3 above)
[the remark this answers is not in the document]
Comment (MarkWoulfeAbly) [reply content/partials/shared/_channel_namespaces.textile:14]:
```suggestion
* **TLS only** - if enabled, only clients who have connected to Ably over TLS will be allowed to use matching channels. "Find out more":https://support.ably.com/solution/articles/3000045208
```

File: content/realtime/channels/channel-parameters/rewind.textile (hunk 4 above)
[the remark this answers is not in the document]
Comment (MarkWoulfeAbly) [reply content/realtime/channels/channel-parameters/rewind.textile:137]:
```suggestion
If you have enabled the "**Persist last message** channel rule,":/general/channel-rules-namespaces/ on the channel in question, you can attach with @rewind=1@ to retrieve the last message. Note that only the last message can be stored long-term (up to a year), and persisting the last message does not work for presence messages.
```

File: content/realtime/versions/v1.1/channel-params.textile (hunk 5 above)
[the remark this answers is not in the document]
Comment (MarkWoulfeAbly) [reply content/realtime/versions/v1.1/channel-params.textile:92]:
```suggestion
If you have enabled the "**Persist last message** channel rule,":/general/channel-rules-namespaces/ on the channel in question, you can attach with @rewind=1@ to retrieve the last message. Note that only the last message can be stored long-term (up to a year), and persisting the last message does not work for presence messages.
```

File: content/partials/shared/_channel_namespaces.textile [hunk 6]
```diff
@@ -1,11 +1,17 @@
-One or more channel namespaces, or channel name prefixes, may be "configured for an app in your dashboard":https://support.ably.com/solution/articles/3000030053-how-do-i-access-my-app-dashboard. When a channel is created whose name is prefixed with one of the namespaces, the channel assumes certain configured attributes associated with that namespace. For example, a channel namespace named "@private@" would match channels named "@private@", "@private:chat@", "@private:chat:mike@".
+One or more channel namespace rules may be "configured for an app in your dashboard":https://support.ably.com/solution/articles/3000030053-how-do-i-access-my-app-dashboard. These are rules which apply to a channel based on its 'namespace'. The namespace is the first colon-delimited segment of its name (from the start up to and including the last character before the @:@). If the channel name contains no colon, the namespace is the entire channel name.
```
[the remark this answers is not in the document]
Comment (MarkWoulfeAbly) [reply content/partials/shared/_channel_namespaces.textile:1]:
```suggestion
One or more channel namespace rules may be "configured for an app in your dashboard":https://support.ably.com/solution/articles/3000030053-how-do-i-access-my-app-dashboard. These are rules which apply to a channel based on its 'namespace'. The namespace is the first colon-delimited segment of its name (from the start, up to and including, the last character before the @:@). If the channel name contains no colon, the namespace is the entire channel name.
```

**Review by kavalerov — APPROVED** [review]
Looks good to me.

File: content/general/channel-rules-namespaces.textile [hunk 7]
```diff
@@ -1,13 +1,7 @@
...
+section: general
 index: 15
 ---
 
-Apps can have one or more channel rules with configurable settings that will be applied to matching channels created in the app.  If no channel rule is found when a channel is created, the default channel rule for the app is applied.
-
-Channel rules allow settings such as whether messages are persisted or TLS required to be configured.  The name in each configured channel rule will match any channel with that name or any channel in that namespace.  For example, a channel rule with the configured namespace of "rss" will match the channel name "rss" as well as the channel name "rss:news" which is a channel within the "rss" namespace.
-
-Each channel rule has the following configurable settings:
-* **Persisted messages** - If enabled, all messages within this namespace will be persisted. You can access stored messages via the History API. The number of hours a message is stored is configurable, "find out more about message persistence":http://support.ably.com/solution/articles/3000030059-how-long-are-messages-stored-for
```
Comment (markcoxsmith) [inline content/general/channel-rules-namespaces.textile:11]: Could the following Hubspot equivalent of the support article be used instead: (https://knowledge.ably.com/how-long-are-messages-stored-for-1)  ? In general we want to phase out the use of Freshdesk support article references as we are moving away from using it

File: content/partials/shared/_channel_namespaces.textile [hunk 8]
```diff
@@ -1,11 +1,17 @@
-One or more channel namespaces, or channel name prefixes, may be "configured for an app in your dashboard":https://support.ably.com/solution/articles/3000030053-how-do-i-access-my-app-dashboard. When a channel is created whose name is prefixed with one of the namespaces, the channel assumes certain configured attributes associated with that namespace. For example, a channel namespace named "@private@" would match channels named "@private@", "@private:chat@", "@private:chat:mike@".
```
Comment (markcoxsmith) [inline content/partials/shared/_channel_namespaces.textile:1]: Could the Freshdesk article be replaced with the following https://knowledge.ably.com/how-do-i-access-my-app-dashboard-1 ?

File: content/general/channel-rules-namespaces.textile (hunk 7 above)
Comment (SimonWoolf) [reply content/general/channel-rules-namespaces.textile:11]: This is in red, meaning it's being deleted.

More generally -- there are 191 links to support articles in the docs repo. They (presumably!) were imported with predictable urls. Surely when we decide to switch over, we will replace the links all at once with a batch find/replace job. I don't think a switchover plan of 'manually replace the links one by one as they turn up in PRs over the next 10 years, and in the meantime have a mixture of the two' is practical.

**Review by markcoxsmith — APPROVED** [review]

File: content/general/channel-rules-namespaces.textile (hunk 7 above)
Comment (markcoxsmith) [reply content/general/channel-rules-namespaces.textile:11]: Thanks Simon and true that will not be ideal and all of the titles of the Hubspot articles are the same as those in Freshdesk so there shouldn't be a need to resort to manually moving everything over

Comment (kavalerov) [reply content/general/channel-rules-namespaces.textile:11]: I am also thinking that we should update this support article - @markcoxsmith can you add section to it that asks about long term persistence of last message? 

Feel free to loop me in if you need help with the copy.

**Review by kavalerov — APPROVED** [review]

**Comment by kavalerov** [comment]
@markcoxsmith can you please merge this, as I don't have merge rights?

**Comment by fliptopbox** [comment]
https://github.com/ably/website/pull/3439 the PR that updates website repo with the updated commit

**Pull request merged**